### PR TITLE
[Core] AssembleSparseMatrixByBlocks added to SparseMatrixMultiplicationUtility

### PR DESCRIPTION
**Pull request by loumalouomega**
This allows to build easily an sparse matrix by its blocks. Moves several methods from MixedULMLinearSolver

### CODE REVIEW

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 1]
```diff
@@ -2076,6 +2023,13 @@ class MixedULMLinearSolver :
 ///@}
 ///@name Type Definitions
 ///@{
+
+// Here one should use the KRATOS_CREATE_LOCAL_FLAG, but it does not play nice with template parameters
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:2027]: could you post the error for future reference?

**Review by philbucher — COMMENTED** [review]
this does not change the existing functionalities of the utility right?

@RiccardoRossi I guess this is up to you, no clue here :)

**Comment by loumalouomega** [comment]
It does not change current behaviour

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 1 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:2027]: Basically is the expansion of the macro which is not designed to template classes

**Comment by loumalouomega** [comment]
This PR does not affect any current thing, except the small refactor mixedulm solver which is extensively tested, and I just moved some methods to the utilities to avoid code repetion. Ping @KratosMultiphysics/technical-committee @RiccardoRossi

**Comment by loumalouomega** [comment]
Ping

File: kratos/utilities/sparse_matrix_multiplication_utility.h [hunk 2]
```diff
@@ -683,6 +683,267 @@ class SparseMatrixMultiplicationUtility
...
+        const DenseMatrix<TMatrix*>& rMatricespBlocks,
+        DenseMatrix<double> ContributionCoefficients = DenseMatrix<double>(0,0),
+        DenseMatrix<bool> TransposeBlocks = DenseMatrix<bool>(0,0)
+        )
+    {
+        const SizeType number_of_rows_blocks = rMatricespBlocks.size1();
+        const SizeType number_of_columns_blocks = rMatricespBlocks.size2();
+
+        // Fill the matrices if they are empty
+        if (ContributionCoefficients.size1() == 0 && ContributionCoefficients.size2() == 0) {
```
Comment (RiccardoRossi) [inline kratos/utilities/sparse_matrix_multiplication_utility.h:703]: what is the expected size of "ContrubtionCoefficients"?

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 3]
```diff
@@ -79,6 +79,12 @@ class MixedULMLinearSolver :
     ///@name Type Definitions
     ///@{
 
+    /// The flag that indicates if the blocks are allocated
+    KRATOS_DEFINE_LOCAL_FLAG( BLOCKS_ARE_ALLOCATED );
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:83]: why this change?

File: kratos/utilities/sparse_matrix_multiplication_utility.h [hunk 4]
```diff
@@ -683,6 +683,267 @@ class SparseMatrixMultiplicationUtility
         }
     }
 
+    /**
+     * @brief This method assembles several sparse matrices into one large sparse matrix
+     * @param rMatricespBlocks The pointers to the matrices we are interested in assemble
+     * @param ContributionCoefficients The matrix containing the coefficients to be considered (copy, so we don't need to provide it)
+     * @param TransposeBlocks The matrix containing the flags telling us to transpose the blocks (copy, so we don't need to provide it)
+     */
+    template <class TMatrix>
```
Comment (RiccardoRossi) [inline kratos/utilities/sparse_matrix_multiplication_utility.h]: are u sure you want to template in TMatrix? anyway to do it efficiently you need to use the internals of the Matrix definition, so i don't think this makes sense

File: kratos/utilities/sparse_matrix_multiplication_utility.h [hunk 5]
```diff
@@ -683,6 +683,267 @@ class SparseMatrixMultiplicationUtility
...
+        }
+        KRATOS_ERROR_IF_NOT(check_non_zero == total_nnz) << "Inconsistent number of non-zero values. Check 1: " << total_nnz << " vs " << check_non_zero << std::endl;
+        KRATOS_ERROR_IF_NOT(nonzero_values == total_nnz) << "Inconsistent number of non-zero values. Check 2: " << total_nnz << " vs " << nonzero_values << std::endl;
+    #endif
+
+        // Initialize matrix with the corresponding non-zero values
+        rMatrix = TMatrix(nrows, ncols, nonzero_values);
+
+        // Fill the new matrix
+        double* Matrix_values = rMatrix.value_data().begin();
```
Comment (RiccardoRossi) [inline kratos/utilities/sparse_matrix_multiplication_utility.h:842]: here you are assuming it is a compressed_matrix...so the use of template for matrix type is effectively not useful

File: kratos/utilities/sparse_matrix_multiplication_utility.h [hunk 6]
```diff
@@ -683,6 +683,267 @@ class SparseMatrixMultiplicationUtility
...
+                                ComputeAuxiliarValuesBlocks<TMatrix>(r_matrix, Matrix_index2, Matrix_values, k, row_end, initial_index_column, ContributionCoefficients(i, j));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Close the matrix
+        rMatrix.set_filled(nrows+1, nonzero_values);
```
Comment (RiccardoRossi) [inline kratos/utilities/sparse_matrix_multiplication_utility.h:880]: also this is specific of compressed_matrix

File: kratos/utilities/sparse_matrix_multiplication_utility.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/sparse_matrix_multiplication_utility.h:703]: The number of blocks in each direction

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 3 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:83]: because In order to reduce code duplication I have taken several methods from mixedulm_linear_solver.h  and moved to sparse_multiplication_utility. I changed also one member variable to a flag because in my opinion is cleaner

File: kratos/utilities/sparse_matrix_multiplication_utility.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/sparse_matrix_multiplication_utility.h]: Basically I am following the design considered in the rest of the sparse_matrix_mutiplication_utilities

File: kratos/utilities/sparse_matrix_multiplication_utility.h (hunk 5 above)
Comment (loumalouomega) [reply kratos/utilities/sparse_matrix_multiplication_utility.h:842]: This is done in the rest of sparse_matrix_multiplication_utility.h. BTW, the builders and solvers assume also that it is a compressed matrix and we still consider templates for the spaces...

File: kratos/utilities/sparse_matrix_multiplication_utility.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/utilities/sparse_matrix_multiplication_utility.h:880]: See, the rest of the sparse_matrix_multiplication_utility.h  is designed for compressed_matrix. This is consistent with what it is in the rest of the utility. Additionally the B&S consider also to work with compressed_matrix and we still consider the space as a template argument. Should we refactor all these implementations first? 

For example: https://github.com/KratosMultiphysics/Kratos/blob/0c0a3dbac08f886faca926d11c0b049206d7cfd8/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L1253 and https://github.com/KratosMultiphysics/Kratos/blob/0c0a3dbac08f886faca926d11c0b049206d7cfd8/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L1280

**Review by RiccardoRossi — COMMENTED** [review]
dear vicente, as i am telling there is no need to template in the matrix so please remove it.

regarding builder and solver you are right in that we are also not taking advantage of the sparse space in there, nevertheless we do take advantage of the dense space, and we use the space for mpi vs openmp parallelization using the same base strategy and schemes so i think it does make sense in there.

In any case B&S is in production by many years and used by everyone so please DO NOT refactor it again

**Comment by loumalouomega** [comment]
Do I change the rest of methods or only the new ones?

**Comment by RiccardoRossi** [comment]
you just added the template no?

**Comment by loumalouomega** [comment]
The old methods also have the same template argument (the ones in the
utility, not in the solver)

**Comment by loumalouomega** [comment]
Done @RiccardoRossi

**Comment by loumalouomega** [comment]
Ping @RiccardoRossi

**Comment by loumalouomega** [comment]
@RiccardoRossi I did what you requested

File: kratos/utilities/sparse_matrix_multiplication_utility.h (hunk 5 above)
Comment (RiccardoRossi) [reply kratos/utilities/sparse_matrix_multiplication_utility.h:842]: as far as i know this is not the case. solvers are used both for ublas and for trilinos...they would not work otherwise

**Review by RiccardoRossi — APPROVED** [review]
